internal/https/server.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2023 - MinIO, Inc. All rights reserved.
+// Use of this source code is governed by the AGPLv3
+// license that can be found in the LICENSE file.
+
+package https
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/minio/kes/internal/fips"
+	"github.com/minio/kes/internal/log"
+)
+
+// Config is a structure containing configuration
+// fields for an HTTPS server.
+type Config struct {
+	// Addr specifies an optional TCP address for the
+	// server to listen on in the form "host:port".
+	// If empty, ":https" (port 443) is used.
+	//
+	// The service names are defined in RFC 6335 and assigned by IANA.
+	// See net.Dial for details of the address format.
+	Addr string
+
+	// Handler handles incoming requests.
+	Handler http.Handler
+
+	// TLSConfig provides the TLS configuration.
+	TLSConfig *tls.Config
+}
+
+// NewServer returns a new HTTPS server from
+// the given config.
+func NewServer(config *Config) *Server {
+	srv := &Server{
+		addr:      config.Addr,
+		tlsConfig: config.TLSConfig,
+	}
+
+	srv.handler = &muxHandler{
+		lock:    srv.lock.RLocker(),
+		Handler: config.Handler,
+	}
+	return srv
+}
+
+// Server is a HTTPS server.
+type Server struct {
+	addr      string
+	handler   *muxHandler
+	tlsConfig *tls.Config
+
+	lock sync.RWMutex
+}
+
+// Update updates the Server's configuration or
+// returns a non-nil error explaining why the
+// server configuration couldn't be updated.
+func (s *Server) Update(config *Config) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if config.Addr != s.addr {
+		return fmt.Errorf("https: failed to update server: '%s' does match existing server address", config.Addr)
+	}
+
+	s.tlsConfig = config.TLSConfig.Clone()
+	s.handler.Handler = config.Handler
+	if s.handler.Handler == nil {
+		s.handler.Handler = http.NewServeMux()
+	}
+	return nil
+}
+
+// UpdateTLS updates the Server's TLS configuration
+// or returns a non-nil error explaining why the
+// server configuration couldn't be updated.
+func (s *Server) UpdateTLS(config *tls.Config) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	s.tlsConfig = config.Clone()
+	return nil
+}
+
+// Start starts the HTTPS server by listening on the
+// Server's address.
+//
+// If the server address is empty, ":https" is used.
+//
+// Start blocks until the given ctx.Done() channel returns.
+// It always returns a non-nil error. Once ctx.Done()
+// returns, the Server gets closed and, if gracefully
+// shutdown, Start returns http.ErrServerClosed.
+func (s *Server) Start(ctx context.Context) error {
+	addr := s.addr
+	if addr == "" {
+		addr = ":https"
+	}
+	listener, err := tls.Listen("tcp", addr, &tls.Config{
+		MinVersion:       tls.VersionTLS12,
+		CipherSuites:     fips.TLSCiphers(),
+		CurvePreferences: fips.TLSCurveIDs(),
+
+		NextProtos: []string{"h2"},
```

Comment:
This is required since:
> HTTP/2 support is only enabled if the Listener returns *tls.Conn connections and they were configured with "h2" in the TLS Config.NextProtos.